File: app/src/main/java/com/ShaHar91/poems/ui/home/categories/adapter/CategoryAdapter.java
package com.shahar91.poems.ui.home.categories.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;

import com.shahar91.poems.R;
import com.shahar91.poems.data.models.Category;
import com.shahar91.poems.ui.base.list.BaseAdapter;
import com.shahar91.poems.ui.base.list.BaseInteractionListener;
import com.shahar91.poems.ui.base.list.BaseViewHolder;

import butterknife.BindView;


public class CategoryAdapter extends BaseAdapter<Category, CategoryAdapter.CategoryInteractionListener, BaseViewHolder<Category, CategoryAdapter.CategoryInteractionListener>> {
    private final CategoryInteractionListener listener;

    public interface CategoryInteractionListener {
        void onCategoryClicked(Category category);
    }

    public CategoryAdapter(Context context, CategoryInteractionListener listener) {
        super(context);

        this.listener = listener;
    }

    @NonNull
    @Override
    public CategoryViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new CategoryViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.list_item_category, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull BaseViewHolder<Category, CategoryInteractionListener> holder, int position) {
        holder.bind(position, getItem(position), listener);
    }

    static class CategoryViewHolder extends BaseViewHolder<Category, CategoryInteractionListener> {
        @BindView(R.id.categoryTv)
        TextView categoryTv;

        CategoryViewHolder(@NonNull View itemView) {
            super(itemView);
        }

        @Override
        public void bind(int position, Category item, CategoryInteractionListener baseInteractionListener) {
            categoryTv.setText(item.getName());

            itemView.setOnClickListener(view -> baseInteractionListener.onCategoryClicked(item));
        }
    }

}

File: app/src/main/java/com/ShaHar91/poems/ui/home/categories/adapter/CategoryAdapter.kt
package com.shahar91.poems.ui.home.categories.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import be.appwise.core.ui.base.list.BaseViewHolder
import com.shahar91.poems.databinding.ListItemCategoryBinding
import com.shahar91.poems.domain.model.Category

class CategoryAdapter(
    private val onCategoryClicked: (category: Category) -> Unit
) : ListAdapter<Category, CategoryAdapter.CategoryViewHolder>(CategoryDiffCallback()) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CategoryViewHolder {
        return CategoryViewHolder(ListItemCategoryBinding.inflate(LayoutInflater.from(parent.context), parent, false))
    }

    override fun onBindViewHolder(holder: CategoryViewHolder, position: Int) {
        val item = getItem(position)
        holder.bind(item)
    }

    inner class CategoryViewHolder(private val binding: ListItemCategoryBinding) :
        BaseViewHolder<Category>(binding.root) {

        override fun bind(item: Category) {
            binding.category = item
            binding.root.setOnClickListener { onCategoryClicked(item) }
        }
    }
}

class CategoryDiffCallback : DiffUtil.ItemCallback<Category>() {
    override fun areItemsTheSame(oldItem: Category, newItem: Category): Boolean {
        return oldItem.id == newItem.id
    }

    override fun areContentsTheSame(oldItem: Category, newItem: Category): Boolean {
        return oldItem == newItem
    }
}